src/main/java/no/ebax/sdk/BaseClient.java

```diff
@@ -0,0 +1,124 @@
+package no.ebax.sdk;
+
+import accesstoken.model.AccessTokenResponse;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.*;
+import java.util.concurrent.Flow;
+import java.util.function.Supplier;
+
+public class BaseClient {
+
+    private Supplier<String> accessTokenSupplier;
+    private ApiClient apiClient;
+
+    public BaseClient(ApiClient apiClient, String apimKey, String username, String password) {
+        this.accessTokenSupplier = new AccessTokenSupplier("/token", apimKey, username, password);
+        this.apiClient = apiClient;
+    }
+
+    public Flow.Publisher<String> post(
+            String uri,
+            String correlationId
+    ) {
+        return post(uri, null, correlationId, Collections.emptyMap());
+    }
+
+    public Flow.Publisher<String> post(
+            String uri,
+            Flow.Publisher<String> body,
+            String correlationId
+    ) {
+        return post(uri, body, correlationId, Collections.emptyMap());
+    }
+
+    public Flow.Publisher<String> post(
+            String uri,
+            Flow.Publisher<String> body,
+            String correlationId,
+            Map<String, List<String>> headers
+    ) {
+        var allHeaders = new HashMap<>(headers);
+        allHeaders.put("X-Correlation-Id", Collections.singletonList(correlationId));
+        allHeaders.put("Authorization", Collections.singletonList("Bearer " + accessTokenSupplier.get()));
+        return apiClient.post(uri, body, allHeaders, String.class);
+    }
+
+    private class AccessTokenSupplier implements Supplier<String>, Flow.Subscriber<AccessTokenResponse> {
+        private final String uri;
+        private final String apimKey;
+        private String username;
+        private String password;
+
+        private String token;
+        private Date expiry;
+
+        private boolean waiting = false;
+
+        public AccessTokenSupplier(String uri, String apimKey, String username, String password) {
+            this.uri = uri;
+            this.apimKey = apimKey;
+            this.username = username;
+            this.password = password;
+        }
+
+        @Override
+        public synchronized String get() {
+            if (isExpired()) fetchNewToken();
+            try {
+                while(waiting) Thread.sleep(100);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            return token;
```

Comment:
You can do without locking and polling. Change the `token` to be an `AtomicReference` and let this method get on that directly. It should validate the token expiration before returning and throw an exception if it is not (alternatively just use the expired token). The `AccessTokenSupplier` should upon init populate and also schedule refresh in the background five minutes prior to expiration. Upon refresh completion the next refresh should be scheduled and so on...

Reply:
This background process also needs to handle transient failures, ie. retry on connection failures, HTTP 500, timeouts...

Reply:
Are you sure continuously refreshing is a good idea? Sounds a bit spammy, although not too problematic either..

Reply:
Not sure how to avoid lags for those unfortunate requests otherwise. The APIs are for system to system integration, so I think it is ok.